packages/styled-components/src/utils/mixinDeep.ts

```diff
@@ -1,60 +1,33 @@
 import { ExtensibleObject } from '../types';
+import isPlainObject from './isPlainObject'
 
-/**
-  mixin-deep; https://github.com/jonschlinkert/mixin-deep
-  Inlined such that it will be consistently transpiled to an IE-compatible syntax.
-
-  The MIT License (MIT)
-
-  Copyright (c) 2014-present, Jon Schlinkert.
-
-  Permission is hereby granted, free of charge, to any person obtaining a copy
-  of this software and associated documentation files (the "Software"), to deal
-  in the Software without restriction, including without limitation the rights
-  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-  copies of the Software, and to permit persons to whom the Software is
-  furnished to do so, subject to the following conditions:
-
-  The above copyright notice and this permission notice shall be included in
-  all copies or substantial portions of the Software.
-
-  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-  THE SOFTWARE.
-*/
-const isObject = (val: any) => {
-  return (
-    typeof val === 'function' || (typeof val === 'object' && val !== null && !Array.isArray(val))
-  );
-};
+function isRecursible(obj: any) {
+  return isPlainObject(obj) || Array.isArray(obj);
+}
 
-const isValidKey = (key: string) => {
-  return key !== '__proto__' && key !== 'constructor' && key !== 'prototype';
-};
+function mixinRecursively(target: any, source: any) {
+  const isRecursive = isRecursible(target);
+  for (const [key, sourceVal] of Object.entries(source)) {
```

Comment:
I'd love to avoid the use of `for-of` and `Object.entries` here. Simply because it's not a rather performant variant and it's imho a bit of unnecessary uses of two non-ES primitives. There's a simpler way here though 👀 

```js
function mixinRecursively(target: any, source: any) {
  if (Array.isArray(target)) {
    for (let key = 0; key < target.length; key++) {
      target[key] = mixinRecursively(target[key], source[key]);
    }
    return target
  } else if (isPlainObject(target)) {
    for (const key in source) {
      target[key] = mixinRecursively(target[key], source[key]);
    }
    return target
  } else if (source != null) {
    return source
  } else {
    return target
  }
}
```

_Note:_ We don't need a `hasOwnProperty` check here I realised, since it's a plain object.

I did notice that we're assigning and mutating here, but seems like we did that before, so 🤷 :+1:
Also, I may have forgotten a case, but simply pointing out here that we can avoid `for-of` and `Object.entries` here entirely without drawbacks :v:

Reply:
Thanks for the suggestion @kitten! I refactored my original implementation to simplify & stop using `Object.entries` and `for..of` as you suggested.   Unfortunately couldn't use your implementation as is.  I added some tests in my latest commit that fail in your implementation (see below).    Also, while testing with your code I realized that `isPlainObject` was actually broken & was returning `true` for some objects with a custom prototype.  I updated the PR and took a stab at fixing that. 

Here is the test output with your implementation:

```
  ● should shallow merge properties if target is not a POJO or array

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `should shallow merge properties if target is not a POJO or array 1`

    - Snapshot  - 6
    + Received  + 0

      MyObj {
    -   "a": Object {
    -     "b": 2,
    -     "d": Object {
    -       "f": "f",
    -     },
    -   },
        "bar": 6,
        "foo": 5,
      }

      47 |   const obj2 = { a: { b: 2, d: { f: 'f' } } };
      48 |
    > 49 |   expect(mixinDeep(obj1, obj2)).toMatchInlineSnapshot(`
         |                                 ^
      50 | MyObj {
      51 |   "a": Object {
      52 |     "b": 2,

      at Object.<anonymous> (packages/styled-components/src/utils/test/mixinDeep.test.ts:49:33)

  ● should properly mixin arrays

    expect(received).toEqual(expected) // deep equality

    - Expected  - 5
    + Received  + 0

      Array [
        12,
    -   22,
    -   Object {
    -     "abc": 123,
    -   },
    -   true,
      ]

      158 |     60,
      159 |   ]);
    > 160 |   expect(mixinDeep([10], [11, 21, { abc: 123 }, true], [12, 22])).toEqual([
          |                                                                   ^
      161 |     12,
      162 |     22,
      163 |     { abc: 123 },

      at Object.<anonymous> (packages/styled-components/src/utils/test/mixinDeep.test.ts:160:67)
```